### PR TITLE
sliding window

### PR DESCRIPTION
**Pull request by ducthangng**
this code can work!

But I have a current issue with nginxt:

`
conns = append(conns, []Conn{
		{
			Title:   "server 1",
			Address: "localhost:8081",
			Conn:    newConn("localhost:8080"),
		},
	}...)
`

with this change, the client call the server via nginx, which is running on port 8080. The nginx configuration can be found here: https://github.com/ducthangng/drl/blob/nginx-issue/nginx.conf

The error from client:

`
2023/12/08 19:58:13 unexpected error:  connection error: desc = "error reading server preface: http2: frame too large"
`

The reason I know this originated in the client is that the server has not received any request yet. This only pop up when I change to nginx, too.

### CODE REVIEW

File: bucket/bucket.go [hunk 1]
```diff
@@ -47,19 +49,51 @@ func (limiter *DistributedRateLimiter) addToken() {
 	limiter.Mutext.Lock()
 	defer limiter.Mutext.Unlock()
```
Comment (lovung) [inline bucket/bucket.go]: this mutex can only protect on current service. How to make sure if you have 1000 services concurrent accessing?

Comment (ducthangng) [reply bucket/bucket.go]: Dilemma 1: Concurrence: A and B takes 1 from redis, -1 and return 0. (at the same time)
Dilemma 2: Dead Server: if one server dies, how do you do?
Dilemma 3: More nodes, less node, die nodes?
Dilemma 4: 1 masters or more masters for each of the rate limiter.
Dilemma 5 (for token-base): control the value added to limiter on the go

---> Synchronize request between servers to the 
 
(alex xu's system design rate limiter chapter 4)

Comment (ducthangng) [reply bucket/bucket.go]: Existed problems:
1. Dilemma of Concurrency is not solved.
2. Database redundance: Zookeeper is a kv storage, so is redis. What should be used? Why

Comment (ducthangng) [reply bucket/bucket.go]: 1. Concurrency can be solved by redis (INCR/DESC).
2. Database redundancy: Zookeeper is a coordination. Redis is the database.

File: service/laptop_store.go [hunk 2]
```diff
@@ -2,33 +2,34 @@ package service
...
-	Find(id string) (*pb.Laptop, error)
+	Save(ctx context.Context, laptop *pb.Laptop) error
+	Find(ctx context.Context, id string) (*pb.Laptop, error)
 	Search(ctx context.Context, filter *pb.Filter, found func(laptop *pb.Laptop) error) error
 }
 
 type InMemoryLaptopStore struct {
-	mutex   sync.RWMutex
-	laptops map[string]*pb.Laptop
+	mutex       sync.RWMutex
```
Comment (lovung) [inline service/laptop_store.go:20]: Mutex for what?

**Pull request merged**